### PR TITLE
Update node contract

### PR DESCRIPTION
**Pull request by neekolas**
### TL;DR

- Rewrites Nodes contract using the ERC721 standard as a base
- Adds a bunch of tests for the new smart contract
- Removes a deprecated flag from our mockery config

## Why ERC721

1. It doesn't seem crazy to model this registry as a NFT, so we might as well use the standard.
2. We get all the OpenZeppelin audited smart contract libraries for free
3. It'll work with the entire rest of the crypto ecosystem (external indexers, other smart contracts, etc) automatically

### AI Assisted Changelog

1. **Smart Contract Changes**:
   - Implemented ERC721 standard in the `Nodes` contract.
   - Added new functions for node management using NFTs: `addNode`, `removeNode`, `updateHttpAddress`, `updateMtlsCert`, and `updateHealth`.
   - Added events `NodeUpdated` and `NodeRemoved` for better tracking of node changes.
   - Restricted node ownership transfer.

2. **Go Bindings Update**:
   - Updated the ABI bindings to reflect new changes in the `Nodes` contract.
   - Added data structures `NodesNode` and `NodesNodeWithId` to handle the new Node data.
   - Implemented necessary function bindings for node management in Go.

3. **Testing**:
   - Added new test cases in Solidity to cover the new node management functionalities.
   - Renamed test contract from `CounterTest` to `GroupMessagesTest` and added `NodesTest` for comprehensive testing of node functionalities.

### CODE REVIEW

**Comment by neekolas** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/106?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#106** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/106?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#105** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/105?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#103** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/103?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#102** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/102?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @neekolas and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: contracts/test/Nodes.sol [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        Nodes.Node memory node = _randomNode(true);
+        address operator = vm.randomAddress();
+
+        uint16 nodeId = nodes.addNode(
+            operator,
+            node.signingKeyPub,
+            node.httpAddress
+        );
+
+        vm.prank(operator);
```
Comment (neekolas) [inline contracts/test/Nodes.sol:122]: `vm.prank` is cool. It just lets you change the `msg.sender` value to any other address for a test, so you can impersonate someone and see how the contract behaves.

File: pkg/registry/registry.go [hunk 2]
```diff
@@ -1,11 +1,10 @@
...
 type Node struct {
-	Index         int
-	PublicKey     []byte
-	GrpcAddress   string
-	DisabledBlock *uint64
-	// Maybe add mTLS cert here
+	NodeId      int
+	SigningKey  []byte
+	HttpAddress string
+	MtlsCert    []byte
```
Comment (neekolas) [inline pkg/registry/registry.go:7]: This field goes away in the next PR

File: contracts/src/Nodes.sol [hunk 3]
```diff
@@ -1,91 +1,162 @@
...
-    Add a node to the network
+    Mint a new node NFT and store the metadata in the smart contract
      */
     function addNode(
-        bytes calldata publicKey,
+        address to,
+        bytes calldata signingKeyPub,
+        string calldata httpAddress
+    ) public onlyOwner returns (uint16) {
+        uint16 nodeId = _nodeIdCounter;
```
Comment (richardhuaaa) [inline contracts/src/Nodes.sol:46]: We should not allow a node to have node ID 0, because an originator SID of 0 means that the originator is the blockchain. Wondering if we should pre-increment _nodeIdCounter rather than post-incrementing it?

File: contracts/src/Nodes.sol [hunk 4]
```diff
@@ -1,91 +1,162 @@
...
+     */
+    function transferFrom(
+        address from,
+        address to,
+        uint256 tokenId
+    ) public override {
+        require(
+            _msgSender() == owner(),
+            "Only the contract owner can transfer Node ownership"
+        );
```
Comment (richardhuaaa) [inline contracts/src/Nodes.sol:69]: Small nit, any reason this one uses a require statement, but `updateHealth` uses `public onlyOwner`?

File: contracts/src/Nodes.sol [hunk 5]
```diff
@@ -1,91 +1,162 @@
...
+     */
+    function updateHttpAddress(
+        uint256 tokenId,
         string calldata httpAddress
-    ) public onlyOwner {
+    ) public {
         require(
-            bytes(nodes[publicKey].httpAddress).length == 0,
-            "Node already exists"
+            _msgSender() == ownerOf(tokenId),
```
Comment (richardhuaaa) [inline contracts/src/Nodes.sol:81]: I wonder if there are tradeoffs to allowing/encouraging nodes to update the HTTP address at any point, rather than going through a more heavyweight process via the DAO? For example, ensuring there is a graceful failover etc. Don't have a super strong opinion here

**Review by richardhuaaa — APPROVED** [review]
Using ERC721 is a smart idea. Will ecosystem tools break if we don't specify base/token URI? I suppose we could also generate some fun images too. Also do we need to block `safeTransferFrom`?

File: contracts/src/Nodes.sol (hunk 3 above)
Comment (neekolas) [reply contracts/src/Nodes.sol:46]: I have a comment somewhere about this. I kinda think we should start at 1000 just to be safe. Would let us partition the smart contracts later if we wanted to.

File: contracts/src/Nodes.sol (hunk 4 above)
Comment (neekolas) [reply contracts/src/Nodes.sol:69]: Only that I wanted a better error message, but agree the inconsistency feels weird

File: contracts/src/Nodes.sol (hunk 5 above)
Comment (neekolas) [reply contracts/src/Nodes.sol:81]: I don't really know. Wondering the same thing. Worth a discussion.

**Review by neekolas — COMMENTED** [review]
> Will ecosystem tools break if we don't specify base/token URI?

The OpenZeppelin default just sets it to "". I would expect tools to handle that, since some generative NFTs are on-chain only. But I'm with you that at some point we'll do an extra-credit project and make a URI that has JSON metadata with a fun image and more info. Then it fully looks like a normal NFT.

> Also do we need to block safeTransferFrom?

`safeTransferFrom` calls `transferFrom` internally, so we get them both with this override.
